float32[] ranges
bool[] valid
bool[] los
